lib/Doctrine/DBAL/Driver/AbstractDriverException.php

```diff
@@ -36,6 +36,42 @@ public function __construct($message, $sqlState = null, $errorCode = null)
         $this->sqlState  = $sqlState;
     }
 
+    /**
+     * @return AbstractDriverException
+     */
+    public static function noInsertId() : self
+    {
+        $message = 'No identity value was generated by the last statement.';
+
+        if (static::class === self::class) {
```

Comment:
Is this method supposed to be called on the abstract class? I'd rather remove this block since it's a temporary solution than complicate it that much.

Another thought I had in mind is, do we really need driver-specific exceptions which are not originated in the underlying drivers? The primary reason for having `OCI8Exception`, `MysqliException`, etc. separated out is that different drivers provide different information about errors and different APIs for obtaining this information.

In the case of exceptions like this which originate in the DBAL, it's not applicable. The information is always the same.

Unlike most other use cases, where this separation is useful when a client is interested in catching only a specific type of exceptions, DBAL clients are not interested in driver-specific exceptions because the driver is abstracted out.

Another point is, an exception originated in DBAL, cannot properly implement `Doctrine\DBAL\Driver\DriverException` because it doesn't have error code and SQLSTATE by definition.

Can we move these new methods to `Doctrine\DBAL\DBALException` (even if it's an API change) and find a better place later?

Reply:
> Is this method supposed to be called on the abstract class? I'd rather remove this block since it's a temporary solution than complicate it that much.

Yes, this is to allow calling `AbstractDriverException::noInsertId()`. There is no other solution to factorize the `noInsertId()` method as you requested it, without either making `AbstractDriverException` not abstract, or creating more driver-specific exception classes.

> Another thought I had in mind is, do we really need driver-specific exceptions which are not originated in the underlying drivers? (...) Unlike most other use cases, where this separation is useful when a client is interested in catching only a specific type of exceptions, DBAL clients are not interested in driver-specific exceptions because the driver is abstracted out.

Exactly. That's why I [suggested earlier](https://github.com/doctrine/dbal/pull/3335#discussion_r234362640) that `DriverException` might not be an interface, but a concrete class.

I get your point that subclasses allow for driver-specific factory methods such as `fromSqlSrvErrors()`, `fromPDOException()` and so on, but they provide nothing that could not be implemented in the `Driver\Connection` classes.

For example, in `SqlSrvConnection`: `throw SQLSrvException::fromSqlSrvErrors()` could be replaced with `throw $this->createExceptionFromSqlSrvErrors()`, that would create a generic `DriverException`. This removes the need from an `SQLSrvException` class altogether.

> Another point is, an exception originated in DBAL, cannot properly implement Doctrine\DBAL\Driver\DriverException because it doesn't have error code and SQLSTATE by definition.

Error code and SQLSTATE are currently optional in `DriverException`, but anyway I'm not sure why you would throw a DriverException from another place of the DBAL?

As I see it, the driver is the lowest layer of DBAL, and the rest of the DBAL is another layer built on top of it: it uses drivers and handles DriverExceptions, but never throws them.

> Can we move these new methods to Doctrine\DBAL\DBALException (even if it's an API change) and find a better place later?

`DBALException` does not implement `DriverException`, so this would not respect the contract!

What I would suggest, if you see no other issue here, is that we merge this PR as is: it provides well encapsulated changes, and has just a handful of well-documented exotic ways to throw exceptions; as soon as this is merged, I can file another PR to propose to change `interface DriverException` to `class DriverException`, and get rid of the driver-specific exceptions. This should be a simple change that will clean up the codebase, and will fix all the temporary hacks introduced here. And we can start a fresh discussion there if issues arise.

What do you think?

Reply:
> For example, in `SqlSrvConnection`: `throw SQLSrvException::fromSqlSrvErrors()` could be replaced with `throw $this->createExceptionFromSqlSrvErrors()`, that would create a generic `DriverException`. This removes the need from an `SQLSrvException` class altogether.

This way, you could throw these exceptions only from connections. They should be also available to statements. That's why they are separate classes.

> `DBALException` does not implement `DriverException`, so this would not respect the contract!

What contract? `Connection::lastInsertId()` doesn't declare any thrown exception. Throwing an exception (which can happen even in master) is already vilation of the contract.

Reply:
> This way, you could throw these exceptions only from connections. They should be also available to statements. That's why they are separate classes.

No problem if implemented as a static method on the Connection, that can be used by the Statement as well.

> What contract? Connection::lastInsertId() doesn't declare any thrown exception. 

[It does](https://github.com/doctrine/dbal/pull/3335/files#diff-7dbc9d70e253495c562b8b2d74f27957R79) now, that's the whole point of what we've been discussing so far, and the contract that the newly introduced tests check.

Reply:
I've pushed a commit in another branch, that I will move to a PR once/if this is merged:
https://github.com/BenMorel/dbal/commit/f9e0c5a55d10a499a57444aeffaab6bd207861cc

It does exactly what I suggested:

- `AbstractDriverException` is now a concrete `DriverException`, replacing the interface
- All sub-exceptions are now gone
- All exotic `new class ... extends AbstractDriverException` are now replaced with `new DriverException`
- Factory methods from sub-exceptions are moved to a `public static` method in the Connection, that can be used from within the Connection itself and from the Statement
- **All relevant Connection methods are now explicitly documented as throwing `DriverException`**

This last point is very important: so far, most of the methods had no documented exceptions, while `query()` and `exec()` were documented as throwing `DBALException`, **a contract that was never respected**:

- `PDOConnection::query()` and `exec()` throw `Driver\PDOException`, which is a `DriverException`, not a `DBALException`
- `DB2Exception::query()` and `exec()` throw `DB2Exception`, which is a `DriverException`
- `SQLSrvException::query()` and `exec()` throw `SQLSrvException` (via `SQLSrvStatement::execute()`), which is a `DriverException`
- ... and so on.

This future PR will make these things right.